File: src/domain/usecases/GetKnowledgeMapUseCase.ts
import { ICourseRepository } from '../repositories/ICourseRepository';
import { ILearnerRepository } from '../repositories/ILearnerRepository';
import { Skill } from '../entities/Skill';
import { SkillScore } from '../entities/SkillScore';
import { SkillScoringEngine } from '../personalization/SkillScoringEngine';

export interface KnowledgeMapData {
  courseId: string;
  skills: Skill[];
  scores: Record<string, SkillScore>;
  weakestSkill: SkillScore | null;
  strongestSkill: SkillScore | null;
  overallScore: number;
}

export class GetKnowledgeMapUseCase {
  constructor(
    private courseRepo: ICourseRepository,
    private learnerRepo: ILearnerRepository
  ) {}

  async execute(courseId: string): Promise<KnowledgeMapData> {
    const skills = await this.courseRepo.getSkillsByCourseId(courseId);
    const profile = await this.learnerRepo.getProfile();
    const scores = profile.scoresByCourse[courseId] || {};

    let weakestSkill: SkillScore | null = null;
    let strongestSkill: SkillScore | null = null;

    Object.values(scores).forEach((s) => {
      if (!weakestSkill || s.score < weakestSkill.score) {
        weakestSkill = s;
      }
      if (!strongestSkill || s.score > strongestSkill.score) {
        strongestSkill = s;
      }
    });

    const overallScore = SkillScoringEngine.computeOverallScore(scores);

    return {
      courseId,
      skills,
      scores,
      weakestSkill,
      strongestSkill,
      overallScore,
    };
  }
}

File: mobile/lib/domain/usecases/get_knowledge_map_use_case.dart
import '../repositories/i_course_repository.dart';
import '../repositories/i_learner_repository.dart';
import '../entities/skill.dart';
import '../entities/skill_score.dart';

class KnowledgeMapData {
  final String courseId;
  final List<Skill> skills;
  final Map<String, SkillScore> scores;
  final SkillScore? weakestSkill;
  final int averageScore;

  const KnowledgeMapData({
    required this.courseId,
    required this.skills,
    required this.scores,
    this.weakestSkill,
    required this.averageScore,
  });
}

class GetKnowledgeMapUseCase {
  final ICourseRepository courseRepository;
  final ILearnerRepository learnerRepository;

  const GetKnowledgeMapUseCase(this.courseRepository, this.learnerRepository);

  Future<KnowledgeMapData> execute(String courseId) async {
    final skills = await courseRepository.getSkillsByCourseId(courseId);
    final profile = await learnerRepository.getProfile();
    final scores = profile.scoresByCourse[courseId] ?? {};

    SkillScore? weakestSkill;
    int totalScore = 0;
    int count = 0;

    for (final s in scores.values) {
      totalScore += s.score;
      count += 1;
      if (weakestSkill == null || s.score < weakestSkill.score) {
        weakestSkill = s;
      }
    }

    return KnowledgeMapData(
      courseId: courseId,
      skills: skills,
      scores: scores,
      weakestSkill: weakestSkill,
      averageScore: count > 0 ? (totalScore / count).round() : 0,
    );
  }
}
